File: OJ/Leetcode/Algorithm/502. IPO.java
class Solution {
    public int findMaximizedCapital(int k, int W, int[] Profits, int[] Capital) {
        PriorityQueue<int[]> q1 = new PriorityQueue<>((a, b) -> (a[0] - a[1]));
        PriorityQueue<int[]> q2 = new PriorityQueue<>((a, b) -> (b[1] - b[0]));

        for(int i = 0; i < Profits.length; i++) {
            q2.add(new int[] {Capital[i], Profits[i]});
        }

        for(int i = 0; i < k; i++)
        {
            while(!q2.isEmpty() && q2.peek()[0] <= W) {
                q1.add(q2.poll());
            }

            if(q1.isEmpty())
                break;

            W += q1.poll()[1];
        }

        return W;
    }
}

File: OJ/Leetcode/Algorithm/502. IPO.cpp
class Solution {
public:
    int findMaximizedCapital(int k, int W, vector<int>& Profits, vector<int>& Capital) {
        auto cmp1 = [](const auto& a, const auto& b){return a.first < b.first;};
        auto cmp2 = [](const auto& a, const auto& b){return a.second > b.second;};
        priority_queue<pair<int, int>, vector<pair<int, int> >, decltype(cmp1)> q1(cmp1);
        priority_queue<pair<int, int>, vector<pair<int, int> >, decltype(cmp2)> q2(cmp2);

        for(int i = 0; i < Capital.size(); i++)
            q2.push(make_pair(Profits[i], Capital[i]));

        for(int i = 0; i < k; i++)
        {
            while(!q2.empty() && q2.top().second <= W)
            {
                q1.push(q2.top());
                q2.pop();
            }
            if(q1.empty())
                break;
            W += q1.top().first;
            q1.pop();
        }
        return W;
    }
};
